src/tools/compiletest/src/runtest.rs

```diff
@@ -1905,13 +1905,14 @@ impl<'test> TestCx<'test> {
         match output_file {
             TargetLocation::ThisFile(path) => {
                 rustc.arg("-o").arg(path);
+                rustc.arg("--out-dir").arg(&self.config.build_base);
```

Comment:
This has the same bug as below - we should find a way to pass `path` to the `ThisFile` branch. It looks like usually `ThisFile(path)` is derived from the directory name:
```
[src/tools/compiletest/src/runtest.rs:1905] output_file = ThisFile(
    "/home/jnelson/rust-lang/rust/build/x86_64-unknown-linux-gnu/test/ui/const-generics/const-argument-non-static-lifetime.full/a",
)
```
So likely we can pass `arg("--out-dir").arg(path.parent())`.